src/linker/implementation.py

```diff
@@ -11,26 +15,51 @@ def __init__(
         implementation_name: str,
         implementation_config: Optional[Dict[str, Any]],
         container_engine: str,
+        environment: Dict[str, Any],
     ):
         self._pipeline_step_name = step_name
         self.name = implementation_name
         self.config = self._format_config(implementation_config)
         self._container_engine = container_engine
+        self._environment = environment
         self._metadata = self._load_metadata()
         self.step_name = self._metadata[self.name]["step"]
+        self.requires_spark = self._metadata[self.name].get("requires_spark", False)
         self._container_full_stem = self._get_container_full_stem()
 
     def __repr__(self) -> str:
         return f"Implementation.{self.step_name}.{self.name}"
 
     def run(
         self,
+        computing_environment: str,
+        session: Optional["drmaa.Session"],
         runner: Callable,
         container_engine: str,
+        step_id: str,
         input_data: List[Path],
         results_dir: Path,
         diagnostics_dir: Path,
     ) -> None:
+        logger.info(f"Running pipeline step ID {step_id}")
+        if self.requires_spark and computing_environment != "local":
+            drmaa = get_slurm_drmaa()
+            spark_master_url, job_id = build_cluster(
+                session=session,
+                environment=self._environment,
+                step_id=step_id,
+                results_dir=results_dir,
+                diagnostics_dir=diagnostics_dir,
+                input_data=input_data,
+                preserve_logs=False,
+            )
+            # Add the spark master url to implementation config
+            if not self.config:
+                self.config = {}
+            self.config["DUMMY_CONTAINER_SPARK_MASTER_URL"] = spark_master_url
+        else:
+            spark_master_url = None
```

Comment:
Is this needed?

Reply:
Whoops, nope, good catch. 

On a not-that-related side note...do you think we should explicitly unset the DUMMY_CONTAINER_SPARK_MASTER_URL variable after the cluster is torn down? I don't think it matters since the env is limited to this python session scope but wonder if there is potential conflicts if we don't.

Reply:
Is that environment variable actually kept for future implementations? I'd say it should only be passed to the one that needs it.

Reply:
Ah I see. I think it's just for the subprocess actually, so it's fine:

https://github.com/ihmeuw/linker/blob/2c927111a1d41e4303f773eba293e10f62b16f52/src/linker/utilities/singularity_utils.py#L47-L53